### ISSUE TITLE
Upgrade Clang based static analysis and linting tools

### ISSUE THREAD

**Issue by feltech** [OpenAssetIO/OpenAssetIO#1390]
## What

Update to the latest feasible `clang-tidy` and `clang-format` tools

## Why

At time of writing GitHub CI is still on v12 when v18 is the latest stable release. Local usage of later `clang-tidy` versions shows that they do find more and genuine issues, as well as fix bugs in older versions, especially false negatives.

```[tasklist]
### Tasks
- [ ] Remove clang-tidy suppressions from external macros e.g. assert, Catch2
```

**Comment by feltech** [comment]
See comments on https://github.com/OpenAssetIO/OpenAssetIO/issues/812 for some initial investigation re. upgrading Clang tooling.

**Comment by feltech** [comment]
Interestingly, I found that I've been running clang-tidy 13 locally (vs. 12 on CI).  When I try clang-tidy 12 I get unknown compiler option errors (`-flto=auto`). My theory why CI works is that even though clang-tidy 12 is used, clang itself is newer ([v13-v15](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md)).  Since I'm using conda locally, I can't have a mismatched clang and clang-tidy version.

One more reason to upgrade the clang tooling.

**Comment by foundry-markf** [comment]
Can you grab them from PyPi instead? 

https://pypi.org/project/clang-tidy/ v19
https://pypi.org/project/clang-format/ v19

**Comment by feltech** [comment]
> Can you grab them from PyPi instead?
> 
> https://pypi.org/project/clang-tidy/ v19 https://pypi.org/project/clang-format/ v19

Yeah, I noticed this last year (https://github.com/OpenAssetIO/OpenAssetIO/issues/812#issuecomment-1458226505), should make adopting them on CI much easier.  The real work is in fixing (or suppressing) all the issues it raises.
